hasher-matcher-actioner/webapp/src/settings/ActionRuleSettingsTab.jsx

```diff
@@ -22,7 +103,155 @@ export default function ActionRuleSettingsTab() {
             </p>
           </Col>
         </Row>
+        <Row>
+          <Col>
+            <Table bordered>
+              <thead>
+                <tr>
+                  <th>
+                    <Button
+                      className="table-action-button"
+                      onClick={() => {
+                        setAdding(true);
+                      }}>
+                      <ion-icon name="add" size="large" />
+                    </Button>
+                  </th>
+                  <th>Name</th>
+                  <th>Labeled As</th>
+                  <th>Not Labeled As</th>
+                  <th>Action</th>
+                </tr>
+              </thead>
+              <tbody>
+                <tr hidden={!adding}>
+                  <td>
+                    <Button
+                      variant="outline-primary"
+                      className="mb-2 table-action-button"
+                      onClick={() => {
+                        const newActionRule = {
+                          name: nameRef.current.value,
+                          must_have_labels: mustHaveLabelsRef.current.value,
+                          must_not_have_labels:
+                            mustNotHaveLabelsRef.current.value,
+                          action_id: actionRef.current.value,
+                        };
```

Comment:
This is highly unorthodox. [Controlled inputs](https://reactjs.org/docs/forms.html#controlled-components) in react would update their parent's state with their current values when things change. 

Using a ref to obtain current state information defeats the purpose of state.

Consider using controlled components.

Reply:
Retaining refs to sibling components is especially dangerous. React internals 'choose' when to update state. It may use requestAnimationFrame to debounce frequent state updates. So your ref may be out dated if it is referenced any where except the component that defines it.

If there is one thing you take from all of this, avoid refs for state. Only use if it you need to do native events like focus().

Reply:
refactored to not use refs.